Design.md

```diff
@@ -0,0 +1,149 @@
+* [CPU](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/cpu.txt):
+    * `cpu.cfs_period_us` (keep default: 100ms)
+    * `cpu.cfs_quota_us`
+        * `cfs_period_us` means one cpu cores worth of usage
+        * `cfs_period_us` / 2 means 50% of one core
+* [Memory](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/memory.txt):
+    * `memory.limit_in_bytes` (max amount of memory to be used, hard limit)
+    * `memory.soft_limit_in_bytes` (set to 75% of `memory.limit_in_bytes`)
+    * `memory.memsw.limit_in_bytes` (set to `memory.limit_in_bytes`)
+* [Block IO](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/blkio-controller.txt): (Device list readable from `/proc/partitions`)
+    * `blkio.throttle.read_bps_device`
+    * `blkio.throttle.write_bps_device`
+    * `blkio.throttle.read_iops_device`
+    * `blkio.throttle.write_iops_device`
+
+In addition to limiting resource use with cgroups, job-runner also isolates jobs from each other using namespaces. A PID namespace is set up to make sure the job cannot kill processes not spawned by it, and to make sure all child processes are terminated together with the actual job. A mount namespace is used to limit process to a subset of the file system, together with [`pivot_root(2)`](https://linux.die.net/man/2/pivot_root) (see [Understanding Containerization By Recreating Docker](https://itnext.io/linux-container-from-scratch-339c3ba0411d), search for pivot_root). Finally, a network namespace is created to limit network access of the jobs. It must be used together with [Virtual ethernet (VETH)](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#veth) and bridge interfaces if jobs should be allowed to communicate between each other. This also means that internet access must be granted separately.
+
+## Communication and auth
+
+Clients communicate via gRPC with a simple protocol: One service with four endpoints. The communication is secured [RusTLS](https://github.com/ctz/rustls) which properly [audited](https://github.com/ctz/rustls/blob/master/audit/TLS-01-report.pdf) and by design supports only modern, secure cipher suites. Only TLS 1.2/1.3 are used, and authentication is done using ECDSA, Ed25519 or RSA. RusTLS will remove support for cipher suites deemed insecure, and simply keeping the version up to date should be sufficient in the future as well. [Tonic](https://github.com/hyperium/tonic) is used to serve gRPC and almost automatically handles protocol buffers, encryption and related concerns.
```

Comment:
> TLS 1.2/1.3 are used

Any reason not to stick to TLS 1.3?

Reply:
It requires more configuration, and probably another 30-60 minutes looking through API docs. TLS 1.2 with sensible cipher suite selection is as secure as TLS 1.3. The main difference is faster handshake, and that doesn't matter much for continuous connections.